### PR TITLE
client API: add read/write rules

### LINKED ISSUE #873: Add new logic to write
Mainly convert from client api format to coredoc format. Give write access `prefix.*`

Write access:
- create write roles/rules/transitions
- create read roles/rules

### PR DESCRIPTION
**Pull request by charleenfei**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Closes #873

### CODE REVIEW

File: documents/coredocument.go [hunk 1]
```diff
@@ -197,6 +202,47 @@ func (cd *CoreDocument) PrepareNewVersion(collaborators []string, documentPrefix
 
 }
 
+// PrepareNewVersion prepares the next version of the CoreDocument
+// if initSalts is true, salts will be generated for new version.
```
Comment (mikiquantum) [inline documents/coredocument.go]: Lets add a TODO that will replace the original `PrepareNewVersion ` very soon.

File: documents/coredocument.go [hunk 2]
```diff
@@ -197,6 +202,47 @@ func (cd *CoreDocument) PrepareNewVersion(collaborators []string, documentPrefix
 
 }
 
+// PrepareNewVersion prepares the next version of the CoreDocument
```
Comment (mikiquantum) [inline documents/coredocument.go]: Let's keep the linting happy though:
```suggestion
// PrepareNewVersion1 prepares the next version of the CoreDocument
```

File: documents/coredocument.go [hunk 3]
```diff
@@ -76,6 +76,11 @@ type CoreDocument struct {
 	Document coredocumentpb.CoreDocument
 }
 
+type Collaborators struct {
```
Comment (mikiquantum) [inline documents/coredocument.go]:
```suggestion
type CollaboratorsAccess struct {
```
At the moment is very simple as we do not care about field access and that is ok.

File: documents/coredocument.go [hunk 4]
```diff
@@ -197,6 +202,47 @@ func (cd *CoreDocument) PrepareNewVersion(collaborators []string, documentPrefix
...
+		SignatureData:      new(coredocumentpb.SignatureData),
+	}
+
+	err = populateVersions(&cdp, &cd.Document)
+	if err != nil {
+		return nil, err
+	}
+
+	ncd := &CoreDocument{Document: cdp}
+	ncd.addCollaboratorsToReadSignRules(rcs)
```
Comment (mikiquantum) [inline documents/coredocument.go]: Unless `ncd.addCollaboratorsToTransitionRules` as well adds the read rules needed, we need to make this call as well with the write collabs:
`ncd.addCollaboratorsToReadSignRules(wcs)`

File: documents/coredocument.go (hunk 3 above)
Comment (charleenfei) [reply documents/coredocument.go]: :+1:

File: documents/coredocument.go [hunk 5]
```diff
@@ -76,6 +76,11 @@ type CoreDocument struct {
 	Document coredocumentpb.CoreDocument
 }
 
+type CollaboratorsAccess struct {
```
Comment (vedhavyas) [inline documents/coredocument.go]: Why do we need this?
Can you comment this pls

Comment (charleenfei) [reply documents/coredocument.go]: we need this to differentiate the types of collaborators we will be adding to the rules, when we integrate the request header as part of the ClientAPI improvements epic.

Comment (vedhavyas) [reply documents/coredocument.go]: Then, shouldn't this be part of the client protobufs instead?

Comment (charleenfei) [reply documents/coredocument.go]: The structures exist in the client protobufs as well, under the message RequestHeader. but we need this structure here in the coredoc to convert from api wire format on the create/update calls to a usable payload for coredoc methods.

Comment (vedhavyas) [reply documents/coredocument.go]: I think we should'nt create these but instead re-use them. Else there will be converters for just this. And these converters will get tricky once we add the more fiels to the write_acls

File: documents/coredocument.go [hunk 6]
```diff
@@ -457,9 +504,82 @@ func (cd *CoreDocument) GetCollaborators(filterIDs ...identity.DID) ([]identity.
...
+	}
+
+	return collabs, nil
+}
+
+// TODO: remove this method with new GetCollaborators
 // getCollaborators returns all the collaborators who belongs to the actions passed.
 func (cd *CoreDocument) getCollaborators(actions ...coredocumentpb.Action) (ids []identity.DID, err error) {
-	findRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
+	findReadRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
```
Comment (vedhavyas) [inline documents/coredocument.go]: this shouldn't return a bool but instead the err itself. Once you return the error, we should return the error to the caller else we will just swallow the error

File: documents/coredocument.go [hunk 7]
```diff
@@ -457,9 +504,82 @@ func (cd *CoreDocument) GetCollaborators(filterIDs ...identity.DID) ([]identity.
...
+			if err != nil {
+				return false
+			}
+			ids = append(ids, did)
+		}
+
+		return false
+	}, actions...)
+
+	if err != nil {
```
Comment (vedhavyas) [inline documents/coredocument.go]: same as above

File: documents/coredocument.go [hunk 8]
```diff
@@ -457,9 +504,82 @@ func (cd *CoreDocument) GetCollaborators(filterIDs ...identity.DID) ([]identity.
...
+		return nil, err
+	}
+
+	return ids, nil
+}
+
+// getWriteCollaborators returns all the collaborators which have access to the transition actions passed in.
+func (cd *CoreDocument) getWriteCollaborators(actions ...coredocumentpb.TransitionAction) (ids []identity.DID, err error) {
+	findTransitionRole(cd.Document, func(_, _ int, role *coredocumentpb.Role) bool {
```
Comment (vedhavyas) [inline documents/coredocument.go]: same as above

File: documents/coredocument.go (hunk 5 above)
Comment (mikiquantum) [reply documents/coredocument.go]: I think we should not expose clientapi definitions into internal methods. This way we can evolve the clientapi independent of the core logic and viceversa.

File: documents/coredocument.go [hunk 9]
```diff
@@ -457,9 +504,82 @@ func (cd *CoreDocument) GetCollaborators(filterIDs ...identity.DID) ([]identity.
...
+
+	rcs, err := cd.getReadCollaborators(coredocumentpb.Action_ACTION_READ_SIGN, coredocumentpb.Action_ACTION_READ)
+	if err != nil {
+		return CollaboratorsAccess{}, err
+	}
+	wcs, err := cd.getWriteCollaborators(coredocumentpb.TransitionAction_TRANSITION_ACTION_EDIT)
+	if err != nil {
+		return CollaboratorsAccess{}, err
+	}
+	rc := filterCollaborators(rcs, filterIDs...)
```
Comment (vimukthi-git) [inline documents/coredocument.go]: these filter calls can be easily avoided by passing a filtered ids to the `getReadCollaborators` and  `getWriteCollaborators`.

Comment (charleenfei) [reply documents/coredocument.go]: those methods don't take filtered ids as a parameter as of now... are you thinking of GetCollaborators?

Comment (vimukthi-git) [reply documents/coredocument.go]: I mean you can modify them to take it as params, I don't know the complexity of changing them, but looks pretty straight forward here?

File: p2p/receiver/handler.go [hunk 10]
```diff
@@ -247,17 +247,17 @@ func (srv *Handler) validateDocumentAccess(ctx context.Context, docReq *p2ppb.Ge
 	switch docReq.AccessType {
 	case p2ppb.AccessType_ACCESS_TYPE_REQUESTER_VERIFICATION:
 		if !m.AccountCanRead(peer) {
-			return errors.New("requester does not have access")
+			return ErrAccessDenied
```
Comment (vimukthi-git) [inline p2p/receiver/handler.go]: 💯

**Pull request merged**